### PR TITLE
refactor(world): add validateNamespace helper

### PR DESCRIPTION
**Pull request by yonadaaa**

### CODE REVIEW

File: packages/world/ts/config/v2/namespaces.ts [hunk 1]
```diff
@@ -19,12 +20,16 @@ export type namespacedTableKeys<world> = world extends { namespaces: infer names
...
+      : ErrorMessage<`\`${key & string}\` is not a valid Namespace config option.`>;
+};
+
 export type validateNamespaces<namespaces, scope extends Scope = AbiTypeScope> = {
-  [namespace in keyof namespaces]: {
-    [key in keyof namespaces[namespace]]: key extends "tables"
-      ? validateTables<namespaces[namespace][key], scope>
-      : namespaces[namespace][key];
-  };
+  [namespace in keyof namespaces]: validateNamespace<namespace, scope>;
```
Comment (holic) [inline packages/world/ts/config/v2/namespaces.ts:32]:
```suggestion
  [namespace in keyof namespaces]: validateNamespace<namespaces[namespace], scope>;
```

File: packages/world/ts/config/v2/namespaces.ts [hunk 2]
```diff
@@ -19,12 +20,16 @@ export type namespacedTableKeys<world> = world extends { namespaces: infer names
     }[keyof namespaces]
   : never;
 
+export type validateNamespace<namespace, scope extends Scope = AbiTypeScope> = {
+  readonly [key in keyof namespace]: key extends "tables"
+    ? validateTables<namespace[key], scope>
+    : key extends keyof NamespaceInput
+      ? conform<namespace[key], NamespaceInput[key]>
+      : ErrorMessage<`\`${key & string}\` is not a valid Namespace config option.`>;
```
Comment (holic) [inline packages/world/ts/config/v2/namespaces.ts:28]:
```suggestion
      : ErrorMessage<`\`${key & string}\` is not a valid namespace config option.`>;
```

File: packages/world/ts/config/v2/worldWithShorthands.ts [hunk 3]
```diff
@@ -54,7 +54,7 @@ export type validateNamespacesWithShorthands<namespaces, scope extends Scope = A
   [namespace in keyof namespaces]: {
     [key in keyof namespaces[namespace]]: key extends "tables"
       ? validateTablesWithShorthands<namespaces[namespace][key], scope>
-      : validateNamespaces<namespaces[namespace], scope>[key];
+      : validateNamespace<namespaces[namespace], scope>[key];
```
Comment (holic) [inline packages/world/ts/config/v2/worldWithShorthands.ts:57]: might be nice to add some regression tests to make sure this doesn't slip through the cracks again!

File: packages/world/ts/config/v2/namespaces.ts [hunk 4]
```diff
@@ -19,12 +20,16 @@ export type namespacedTableKeys<world> = world extends { namespaces: infer names
     }[keyof namespaces]
   : never;
 
+export type validateNamespace<namespace, scope extends Scope = AbiTypeScope> = {
```
Comment (holic) [inline packages/world/ts/config/v2/namespaces.ts:23]: we ought to have the corresponding `validateNamespace` runtime function as well

Comment (yonadaaa) [reply packages/world/ts/config/v2/namespaces.ts:23]: yess that would be neater

File: packages/world/ts/config/v2/worldWithShorthands.ts (hunk 3 above)
Comment (yonadaaa) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: Agreed!

Comment (yonadaaa) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: Agreed!

Would we just test that each namespace in a shorthand config is valid?

Comment (holic) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: I think a type+runtime test for a valid+invalid namespace should be good

Comment (yonadaaa) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: Only thing is we can't test until `namespaces` is enabled in https://github.com/latticexyz/mud/pull/2880

Comment (holic) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: why? there's already tests for namespaces in `worldWithShorthands.test.ts`

Comment (yonadaaa) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: Doesn't that mean that the `namespaces` key is enabled on `defineWorldWithShorthands`? That feels incorrect

Comment (holic) [inline packages/world/ts/config/v2/worldWithShorthands.ts:57]: need to remember to add this to `validateWorld` too

**Review by holic — APPROVED** [review]

File: packages/world/ts/config/v2/worldWithShorthands.ts (hunk 3 above)
Comment (yonadaaa) [reply packages/world/ts/config/v2/worldWithShorthands.ts:57]: will do that in [yonadaaa/fix-tables-without-namespace](https://github.com/latticexyz/mud/pull/2880)

**Pull request merged**